project-base/app/src/Controller/Front/SocialNetworkController.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Controller\Front;
+
+use App\Model\Customer\User\RegistrationDataFactory;
+use App\Model\Customer\User\RegistrationFacade;
+use Shopsys\FrameworkBundle\Model\Security\Authenticator;
+use Shopsys\FrameworkBundle\Model\SocialNetwork\Exception\SocialNetworkLoginException;
+use Shopsys\FrameworkBundle\Model\SocialNetwork\SocialNetworkConfigFactory;
+use Shopsys\FrameworkBundle\Model\SocialNetwork\SocialNetworkFacade;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+class SocialNetworkController extends AbstractController
+{
+    protected const string REFERER_URL = 'refererUrl';
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Security\Authenticator $authenticator
+     * @param \Shopsys\FrameworkBundle\Model\SocialNetwork\SocialNetworkConfigFactory $socialNetworkConfigFactory
+     * @param \App\Model\Customer\User\RegistrationFacade $registrationFacade
+     * @param \App\Model\Customer\User\RegistrationDataFactory $registrationDataFactory
+     * @param \Shopsys\FrameworkBundle\Model\SocialNetwork\SocialNetworkFacade $socialNetworkFacade
+     */
+    public function __construct(
+        protected readonly Authenticator $authenticator,
+        protected readonly SocialNetworkConfigFactory $socialNetworkConfigFactory,
+        protected readonly RegistrationFacade $registrationFacade,
+        protected readonly RegistrationDataFactory $registrationDataFactory,
+        protected readonly SocialNetworkFacade $socialNetworkFacade,
+    ) {
+    }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param string $type
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function loginAction(Request $request, string $type): Response
+    {
+        // we need to save referer url because login to social network does redirect and after that, we would lose referer url
+        if ($request->getSession()->has(self::REFERER_URL) === false) {
+            $request->getSession()->set(self::REFERER_URL, $request->server->get('HTTP_REFERER'));
+        }
+
+        try {
+            $redirectUrl = $this->generateUrl('front_social_network_login', ['type' => $type], UrlGeneratorInterface::ABSOLUTE_URL);
+            $tokens = $this->socialNetworkFacade->login($type, $redirectUrl);
+
+            return $this->render('Admin/Content/Login/loginAsCustomerUser.html.twig', [
+                'tokens' => $tokens,
+                'url' => $this->getRefererUrl($request, false),
+            ]);
+        } catch (SocialNetworkLoginException $exception) {
+            return $this->redirect($this->getRefererUrl($request, true));
```

Comment:
this is something that was agreed on with a FE dev? Normally, we return UserErrors from API so the FE is able to handle the errors. I see this is not using API at all so this looks like a new approach to the error handling. I would like to discuss it please

Reply:
yep, it's agreed with FE devs, look their commits 😊